dbgd.c

```diff
@@ -298,3 +300,65 @@ static int dbgd_show_front_screen(Dbg__Request *req, Dbg__Response *res)
 
     return DBG__RESPONSE__TYPE__OK;
 }
+
+static int dbgd_call(Dbg__Request *req, Dbg__Response *res)
+{
+    if (!req->has_address)
+        return DBG__RESPONSE__TYPE__ERROR_INCOMPLETE_REQUEST;
+
+    // These variables will be used as parameters for inline assembly.
+    // We make them static as the stack pointer will be changed.
+    // So we want to address them globally.
+    static uint32_t stack_pointer;
+    static uint32_t stack_backup;
+    static uint32_t address;
+
+    // Allocate a stack for working and space for supplied data
+    size_t stack_size = 0x1000;
+    if (req->has_data) {
+      stack_size += req->data.len;
+    }
+    uint8_t* stack_data = malloc(stack_size);
+
+    // Push optional stack contents, starting at top of stack
+    stack_pointer = (uint32_t)&stack_data[stack_size];
+    if (req->has_data) {
+      stack_pointer -= req->data.len;
+      memcpy((void*)stack_pointer, req->data.data, req->data.len);
+    }
+
+    // Set address to call
+    address = req->address;
+
+    asm("pusha\n"
+        "mov %%esp, %[stack_backup]\n" // Keep copy of original stack
+
+        "mov %[stack_pointer], %%esp\n"
+
+        "call *%[address]\n" // Call the function
+
+        "mov %[stack_pointer], %%esp\n" // push all register values to stack
+        "pusha\n"
+
+        "mov %[stack_backup], %%esp\n" // Recover original stack
+        "popa\n"
+
+        : // No outputs
+        : [stack_pointer] "m" (stack_pointer),
+          [stack_backup]  "m" (stack_backup),
+          [address]       "m" (address)
+        : "memory");
+
+    // Get transfer buffer for a set of registers from pusha
+    res->data.len  = 32;
+    res->data.data = get_transfer_buffer(res->data.len);
+    res->has_data = 1;
+
+    // Copy pusha data into return buffer
+    stack_pointer -= res->data.len;
+    memcpy(res->data.data, (void*)stack_pointer, res->data.len);
```

Comment:
Instead of transferring these data back to client, we could instead keep them in a temporary buffer and let the client read them out if it really wants them (higher cost). See my above suggestion regarding having client pass in the (client allocated) stack frame. We could simply pusha to this frame and client can read them out on demand.

One register we could return here however would be `EAX` since that register is used extremely often to pass return values.